### PR TITLE
implement iterate_upper_bound for txn iterators

### PR DESCRIPTION
**Pull request by miasantreble**
Current implementation of current_over_upper_bound_ fails to take into consideration that keys might be invalid in either base iterator or delta iterator. Calling key() in such scenario will lead to assertion failure and runtime errors.
This PR addresses the bug by adding check for valid keys before calling IsOverUpperBound(), also added test coverage for iterate_upper_bound usage in BaseDeltaIterator
(trying to land PR https://github.com/facebook/rocksdb/pull/4702 once again)
Test plan: 
* make check
* run mysqltest.sh

### CODE REVIEW

**Comment by siying** [comment]
Did we understand why the randomized tests failed to catch the bug exposed by MyRocks now?

**Comment by miasantreble** [comment]
I think it's because inside the infinite loop, the update to `current_at_base_` may be delayed, so it's not more reliable to call key() to get the most up to date key. this commit contains the fix https://github.com/facebook/rocksdb/pull/5105/commits/848036d3f120e9cbd2828d32926ccb508fe14a96 and got verified in mysqltest

**Comment by siying** [comment]
@miasantreble the question is, why didn't the randomized test catch it?

File: utilities/write_batch_with_index/write_batch_with_index.cc [hunk 1]
```diff
@@ -264,32 +276,38 @@ class BaseDeltaIterator : public Iterator {
       } else if (!delta_iterator_->status().ok()) {
         // Expose the error status and stop.
         current_at_base_ = false;
-        return;
+        break;
```
Comment (siying) [inline utilities/write_batch_with_index/write_batch_with_index.cc]: It's safer to just return here. I am worried about by mistake we access the data while status is not OK. It always makes things hard to reason about.

File: utilities/write_batch_with_index/write_batch_with_index.cc [hunk 2]
```diff
@@ -264,32 +276,38 @@ class BaseDeltaIterator : public Iterator {
...
       }
       equal_keys_ = false;
       if (!BaseValid()) {
         if (!base_iterator_->status().ok()) {
           // Expose the error status and stop.
           current_at_base_ = true;
-          return;
+          break;
```
Comment (siying) [inline utilities/write_batch_with_index/write_batch_with_index.cc]: Same here. Safer to just return.

File: utilities/write_batch_with_index/write_batch_with_index.cc [hunk 3]
```diff
@@ -264,32 +276,38 @@ class BaseDeltaIterator : public Iterator {
...
         }
 
         // Base has finished.
         if (!DeltaValid()) {
           // Finished
-          return;
+          break;
```
Comment (siying) [inline utilities/write_batch_with_index/write_batch_with_index.cc]: Both of delta and base are at the end. A good idea to return.

File: utilities/write_batch_with_index/write_batch_with_index.cc [hunk 4]
```diff
@@ -264,32 +276,38 @@ class BaseDeltaIterator : public Iterator {
...
         }
         if (delta_entry.type == kDeleteRecord ||
             delta_entry.type == kSingleDeleteRecord) {
           AdvanceDelta();
+          // If the new Delta is valid and >= iterate_upper_bound, stop
+          current_over_upper_bound_ =
+              DeltaValid() && IsOverUpperBound(delta_iterator_->Entry().key);
+          if (current_over_upper_bound_) {
+            return;
+          }
```
Comment (siying) [inline utilities/write_batch_with_index/write_batch_with_index.cc]: Do we consider to do the check inside `AdvanceDelta()` instead?

**Comment by miasantreble** [comment]
ran mysqltest using latest commit 495e6c7 and no crash found

**Comment by siying** [comment]
@miasantreble I think I asked the question twice,  but why didn't the randomized test catch the bug we found in MyRocks test? Or would it catch it?

**Comment by siying** [comment]
Do you consider to move the check to AdvanceDelta() and leaves UpdateCurrent() alone?

File: utilities/write_batch_with_index/write_batch_with_index.cc (hunk 4 above)
Comment (miasantreble) [reply utilities/write_batch_with_index/write_batch_with_index.cc]: `delta_iterator_` modification comes from two paths: 
1) variants of Seek, rely on `UpdateCurrent` to refresh `current_over_upper_bound_`
2) Next/Prev, refresh `current_over_upper_bound_` through `AdvanceDelta` 
If we keep the current code structure, we cannot remove the check from `UpdateCurrent()`

**Pull request closed without merging**